freestyle-effects/shared/src/main/scala/effects/validation.scala

```diff
@@ -0,0 +1,63 @@
+package freestyle.effects
+
+import freestyle._
+import cats.{MonadState}
+import cats.data.{State, Validated, ValidatedNel, NonEmptyList}
+
+
+object validation {
+  /** A validation exception with an explanation. **/
+  trait ValidationException {
```

Comment:
We may consider here wrapping the entire Algebra in a provider similar to how the `ReaderM` and `StateM` effects work so that `ValidationException` can be any `E` provided by the user. The issue with us providing a unsealed datatype which people need to extend from is that once they unwrap the errors and wish to coalesce errors into something else the compiler won't be able to do exhaustive checking that all cases are contemplated.
For example at evaluation time:
```scala
program[ValidationM.Op].exec[Logger].run(List.empty) map { case (errors, a) => 
   errors match {
     case Nil => ...
     case h :: _ => h match {
        case CustomExceptionExtendingValidationException =>
        case _ =>  // Requiered for exahustive matching because ValidationException is open to extension
     }
   }
}
```